Add risky orders push notifications

joiggama opened the pull request:
- [x] ~~Detect risky payments based on spree order risky detection logic~~
- [x] Add tests
- [x] Detect risky orders hooking into the state transition to complete
- [x] Update README to add push notifications configuration instructions.

joiggama committed:
Add risky orders push notification

joiggama commented on app/models/spree/payment_decorator.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def send_risk_notification
+    SpreeSam::Notifications.build(:parse,
+      channels: (ENV["PARSE_CHANNELS"] || []).split(","),
```
Not so sure about this parse `channels` approach.

softr8 replied:
You can leverage Spree::Config object

softr8 commented on app/models/spree/payment_decorator.rb:
```diff
@@ -0,0 +1,29 @@
+  after_save :send_risk_notification, if: :is_risky?
+
+  def is_risky?
+    is_cvv_risky? || is_avs_risky? || state == "failed"
```
what if the credit card gets rejected by payment gateway? it'd change to 'failed' sending the notification, doesn't it?

joiggama replied:
Yes, according to [`Spree::Order#is_risky?`](https://github.com/spree/spree/blob/3-0-stable/core/app/models/spree/order.rb#L499-L501), orders riskiness is defined by [`Spree::Payment.risky`](https://github.com/spree/spree/blob/3-0-stable/core/app/models/spree/payment.rb#L59) scope which considers the `failed` state as a condition to make the payment, and order risky.

softr8 commented on app/models/spree/payment_decorator.rb:
```diff
@@ -0,0 +1,29 @@
+  def send_risk_notification
+    SpreeSam::Notifications.build(:parse,
+      channels: (ENV["PARSE_CHANNELS"] || []).split(","),
+      data: {
```
I feel this object should go in Provider::Parse implementation

joiggama committed:
Simplify payment risk notification building

Remove parse specific arguments, and create notification object based on
the new SpreeSam::Notifications::Providers::Base initialization
signature.

joiggama committed:
Add coveralls integration

joiggama committed:
Only run coverage on demand with COVERAGE env var

joiggama committed:
Refactor notification channels backend

- Improve notification initialization arguments naming
- Generalize notification channels preference (parse might not be the
  only provider using the concept)

joiggama committed:
Hook into order transition to complete for risk detection

joiggama committed:
Add pry-byebug for development debugging

joiggama commented:
@softr8 Updated to hook into order transition to complete.

softr8 commented on app/models/spree/order_decorator.rb:
```diff
@@ -21,4 +21,24 @@
+  private
+
+  def send_risk_notification
+    SpreeSam::Notifications.build(:parse,
```
is :parse the backend? if so, what if it's moved to the configurator so you can switch easily between pusher services

joiggama replied:
Yes, that makes a lot of sense.

softr8 commented:
Besides my comment, LGTM!

joiggama committed:
Add notification provider preference, defaults to "parse"

joiggama commented:
> Besides my comment, LGTM!

Done, made the notification provider a preference that defaults to parse provider.

joiggama committed:
Update README

joiggama committed:
Update LICENSE author
